chia-keys/src/public_key.rs

```diff
@@ -0,0 +1,109 @@
+use crate::derivable_key::DerivableKey;
+use bls12_381_plus::{G1Affine, G1Projective, Scalar};
+use group::Curve;
+use num_bigint::BigUint;
+use sha2::{Digest, Sha256};
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct PublicKey(pub G1Projective);
+
+impl PublicKey {
+    pub fn from_bytes(bytes: &[u8; 48]) -> Option<PublicKey> {
+        G1Affine::from_compressed(bytes)
+            .map(|p| Self(G1Projective::from(&p)))
+            .into()
+    }
+
+    pub fn to_bytes(&self) -> [u8; 48] {
+        self.0.to_affine().to_compressed()
+    }
+
+    pub fn is_valid(&self) -> bool {
+        self.0.is_identity().unwrap_u8() == 0 && self.0.is_on_curve().unwrap_u8() == 1
+    }
+}
+
+impl DerivableKey for PublicKey {
+    fn derive_unhardened(&self, idx: u32) -> Self {
+        let mut hasher = Sha256::new();
+        hasher.update(self.to_bytes());
+        hasher.update(idx.to_be_bytes());
+        let digest = hasher.finalize();
+
+        // in an ideal world, we would not need to reach for the sledge-hammer of
+        // num-bigint here. This would most likely be faster if implemented in
+        // Scalar directly.
+
+        // interpret the hash as an unsigned big-endian number
+        let mut nounce = BigUint::from_bytes_be(digest.as_slice());
+
+        let q = BigUint::from_bytes_be(&[
```

Comment:
Consider `lazy_static` (as in `clvm_rs`) or `once_cell` here to avoid calling this function multiple times.

Reply:
using a lazy_static caused a performance regression.
```
     Running benches/derive_key.rs (/Users/arvid/Documents/dev/chia_rs/target/release/deps/derive_key-7791798e019a8a44)
secret key, unhardened  time:   [323.28 µs 326.14 µs 329.63 µs]
                        change: [+2.8152% +3.9835% +5.4750%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

One reason might be that the `%=` operator is only implemented for where the right-hand side is moved into it. So even with a constant `Q`, I still ave to `clone()` it when calling `%=`.